service/storage_service.cc

```diff
@@ -2871,11 +2871,14 @@ future<> storage_service::handle_state_normal(inet_address endpoint, gms::permit
     if (tmptr->is_normal_token_owner(endpoint)) {
         slogger.info("Node {} state jump to normal", endpoint);
     }
-    std::unordered_set<inet_address> endpoints_to_remove;
+    std::unordered_map<inet_address, bool> endpoints_to_remove;
 
-    auto do_remove_node = [&] (gms::inet_address node) {
+    auto do_remove_node = [&] (gms::inet_address node, bool force = false) {
         tmptr->remove_endpoint(node);
-        endpoints_to_remove.insert(node);
+        auto [it, inserted] = endpoints_to_remove.try_emplace(node, force);
```

Comment:
This is a hacky way to solve the problem. I think we should not use the force way to remove the node but make the old entry not part of the ring and make gossip to remove the old entry the same way as we did for any node that is removed from the cluster. 

You mentioned the following in the commit log:

> If the new node that changed its ip address is decommissioned,
the previous node seems as a normal token owner, just in shutdown
status, but it is not longer in the cluster.

Please give concrete example what is the exact case you want to solve.

So, you meant the following?

n1 was 127.0.0.1 with hostid a-b-c-d
n2 knows n1 was with 127.0.01 but n2 was shutdown
n1 changed the ip to 127.0.0.10 with hostiid a-b-c-d
n1 was removed from the cluster without n2 knowing n1 was removed. 
when n2 was started again, it advertised n1 with ip 127.0.0.1 into the cluster. 

Now,  you want to remove the node with ip 127.0.0.1 in gossip?

The old entry should not be considered as part of the ring because we have the hostid of the old entry, we know the hostid is removed from the cluster, even if we have the old entry and its status is shutdown. 

If the above is not what you meant, please give exact examples.

Reply:
#13775 provides a clear example.
The problem starts when a node changes its ip address and the previous address lingers in gossip in shutdown state:
n1 was 127.0.0.1 with hostid a-b-c-d
n1 is stopped (and the endpoint state for 127.0.0.1 is set to shutdown status)
n2 knows n1 was with 127.0.0.1 and n1 was shutdown
n1 restarts with address 127.0.0.10 with hostid a-b-c-d
n1 is shut down
n2 has two endpoint states with hostid a-b-c-d: 127.0.0.1 and 127.0.0.10, both are in shutdown status
n3 is started, replacing hostid a-b-c-d.  although 127.0.0.1 owns no more tokens, its endpoint_state lingers in gossip and gets in the way.

Reply:
> This is a hacky way to solve the problem. I think we should not use the force way to remove the node but make the old entry not part of the ring

What to you mean by "not part of the ring"?
The old ip address own no tokens after the node changed ip, but it still has an endpoint_state.

> and make gossip to remove the old entry the same way as we did for any node that is removed from the cluster.

But the node is not removed from the cluster.
Only it's previous ip address need to be gone

Reply:
> https://github.com/scylladb/scylladb/issues/13775 provides a clear example.
The problem starts when a node changes its ip address and the previous address lingers in gossip in shutdown state:
n1 was 127.0.0.1 with hostid a-b-c-d
n1 is stopped (and the endpoint state for 127.0.0.1 is set to shutdown status)
n2 knows n1 was with 127.0.0.1 and n1 was shutdown
n1 restarts with address 127.0.0.10 with hostid a-b-c-d
n1 is shut down
n2 has two endpoint states with hostid a-b-c-d: 127.0.0.1 and 127.0.0.10, both are in shutdown status
n3 is started, replacing hostid a-b-c-d. although 127.0.0.1 owns no more tokens, its endpoint_state lingers in gossip and gets in the way.

You mentioned decommission in commit log but the example above has no decommission.

Reply:
> https://github.com/scylladb/scylladb/issues/13775 provides a clear example.
The problem starts when a node changes its ip address and the previous address lingers in gossip in shutdown state:
n1 was 127.0.0.1 with hostid a-b-c-d
n1 is stopped (and the endpoint state for 127.0.0.1 is set to shutdown status)
n2 knows n1 was with 127.0.0.1 and n1 was shutdown
n1 restarts with address 127.0.0.10 with hostid a-b-c-d
n1 is shut down
n2 has two endpoint states with hostid a-b-c-d: 127.0.0.1 and 127.0.0.10, both are in shutdown status
n3 is started, replacing hostid a-b-c-d. although 127.0.0.1 owns no more tokens, its endpoint_state lingers in gossip and gets in the way.

Let's look at this example.  When n3 replaces the node hostid a-b-c-d, since we know the entry 127.0.0.1 has the same hostid a-b-c-d, we can mark the node not part of the ring since it has been replace by another node, e.g., mark it as LEFT.

This way, the entry 127.0.0.1 will be remove from gossip automatically.  No need to hack the force remove endpoint.

Reply:
> You mentioned decommission in commit log but the example above has no decommission.

True. Historically, the motivation was around #14468, but that issue was fixed in a different way by #14507.

Reply:
> Let's look at this example. When n3 replaces the node hostid a-b-c-d, since we know the entry 127.0.0.1 has the same hostid a-b-c-d, we can mark the node not part of the ring since it has been replace by another node, e.g., mark it as LEFT.
>
> This way, the entry 127.0.0.1 will be remove from gossip automatically. No need to hack the force remove endpoint.

I think it's a bad idea, since the node, based on it host_if, has not LEFT the cluster.
We'd like to move away from associating the endpoint state with the ip address and towards associating the endpoint_state with the host_id.  So marking the old ip address as LEFT is futile and quite confusing.

(and BTW, the automatic eventual removal from gossip could be problematic in other ways, like #14991)

I believe that my pending patches for https://github.com/scylladb/scylladb/issues/12282 will solve this by separating the endpoint_state for the old ip address from the new ip address in a better way, returning the updated endpoint_state when searching the host_id. But still, if we know a particular node with a given host_id changed its ip address, we should evict the endpoint_state associated with the old ip address immediately as it's stale.  Why do we need to keep it around?

Reply:
The replacing node uses a new host_id which is different from the node to be replaced now days. Why "since the node, based on its host_id, has not LEFT the cluster"?

Reply:
>  But still, if we know a particular node with a given host_id changed its ip address, we should evict the endpoint_state associated with the old ip address immediately as it's stale. Why do we need to keep it around?

I think you are talking about another issue: changing ip address. Not the one in the example above where the n3 replaces 127.0.0.1 and 127.0.0.10. 

We do not want to keep it but we'd better have a unified way to get rid of them. Look at the force flag, it is very hard to reason about. 

In general, we have two ways to handle the old states: 1) remove it from all nodes immediately 2) set the states correctly immediately, so even if we remove it later, it will not cause only problem.  

Recall the topology states you added recently. A replaced node is classified as left status. Perhaps we could do similar things for node that changes ip address.

Reply:
> > But still, if we know a particular node with a given host_id changed its ip address, we should evict the endpoint_state associated with the old ip address immediately as it's stale. Why do we need to keep it around?
> 
> I think you are talking about another issue: changing ip address. Not the one in the example above where the n3 replaces 127.0.0.1 and 127.0.0.10.

Right.

> 
> We do not want to keep it but we'd better have a unified way to get rid of them. Look at the force flag, it is very hard to reason about.
> 
> In general, we have two ways to handle the old states: 1) remove it from all nodes immediately 2) set the states correctly immediately, so even if we remove it later, it will not cause only problem.
> 
> Recall the topology states you added recently. A replaced node is classified as left status. Perhaps we could do similar things for node that changes ip address.

The problem is that today there is no representation of the node as an entity in gossip.
There is an `endpoint`, which is identified by its inet_address, but it's "the node" that changed its ip address; otherwise it would have only one entry in the _endpoint_state_map, indexed by its host_id.
That's exactly where I'm aiming.

Reply:
Using host_id to index the _endpoint_state_map is going to help the case 127.0.0.1 and 127.0.0.10 shares the same hostid after ip change. But even with ip to index _endpoint_state_map. Current code can already remove  the _endpoint_state_map[127.0.0.1] after ip address changes to 127.0.0.10.  

What you did in the patch is just to make gossip remove 127.0.0.1 faster by calling gossiper.evict_from_membership in addition to gossiper.remove_endpoint. But this is NOT fixing the root cause, because any issue due to accessing of the old 127.0.0.1 entry can still happen before you remove it ( even if it is now faster to remove it).

What I would suggest is to 1) fix any code still uses gossip status to decide if the node is part of the ring. This way even if the old node 127.0.0.1 is still in gossip with SHUTDOWN status it should not cause any problem. we need to figure out  what was going on exactly in the test that fails and why. 2) mark the status of the old node 127.0.0.1 from NORMAL or SHUTDOWN to something else immediately when the hostid collision is detected in storage_service::handle_state_normal.